### ISSUE TITLE
CodonAdaptationIndex math domain error

### ISSUE THREAD

**Issue by conmeehan** [biopython/biopython#243]
I get an error on certain sequences when running the CodonAdaptationIndex module from Bio.SeqUtils.CodonUsage. The error is thrown when using the cai_for_gene method. The error is:

```
File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/Bio/SeqUtils/CodonUsage.py", line 119, in cai_for_gene
    caiValue += math.log(self.index[codon])
ValueError: math domain error
```

I have figured out that the issue is when a codon in the tested sequence. If the index has the value of a codon as 0.00 and the tested sequence has that codon in it, then the error is thrown due to the math.log(0.00). E.g. if in the generated index codon CTA is 0.00 and CTA is present in the tested sequence then the error is thrown. Modifying the code of CodonUsage.py to be:

```
if codon!='ATG' and codon!= 'TGG': #these two codons are always one, exclude them.
                    if self.index[codon]!=0.00: #if the codon is 0 in the index an exception will be thrown so increment the length but do no increase the caiValue
                        caiValue += math.log(self.index[codon])
                    LengthForCai += 1
```

appears to fix the issue but I am unsure if this is the correct way to fix it.

Hope this helps.
Cheers,
Conor

**Comment by Rakksor** [comment]
The [original paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC340524) describing the codon adaptation index says this:

> Note that if a certain codon is never used in the reference set then the CAI for any other gene in which that codon appears becomes zero.To overcome this problem we assign a value of 0.5 to any X<sub>ij</sub> that would otherwise be zero.

So I think the correct way to fix this would be to add the following two lines to the `generate_index` function:
```
            # now generate the index W=RCSUi/RCSUmax:
            rcsu_max = max(rcsu)
            for codon_index, codon in enumerate(codons):
                self.index[codon] = rcsu[codon_index] / rcsu_max

+               if self.index[codon] == 0:
+                   self.index[codon] = 0.5
```

**Comment by tsp-kucbd** [comment]
In line with the original paper, we solved it by adding following line at the end of the `_count_codon` method
`self.codon_count = {c:n if n else 0.5 for c,n in self.codon_count.items()}`

**Comment by peterjc** [comment]
@etal it sounds like this 0.5 fix should be adopted as per the paper?

**Comment by mdehoon** [comment]
Fixed by #4085
